File: app/enums.py
"""Shared enums for the URL shortener application.

This module defines all status and state enums used across the codebase.
Using enums instead of string literals provides type safety and prevents typos.
"""

from enum import StrEnum

__all__ = ["HealthStatus", "ServiceStatus"]


class HealthStatus(StrEnum):
    """Health check status values."""
    
    HEALTHY = "healthy"
    UNHEALTHY = "unhealthy"
    
    @classmethod
    def from_str(cls, value: str) -> "HealthStatus":
        """Safely parse from string, falling back to UNHEALTHY for unknown values."""
        try:
            return cls(value)
        except ValueError:
            return cls.UNHEALTHY


class ServiceStatus(StrEnum):
    """Service operational status values."""
    
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    
    @classmethod
    def from_str(cls, value: str) -> "ServiceStatus":
        """Safely parse from string, falling back to FAILED for unknown values."""
        try:
            return cls(value)
        except ValueError:
            return cls.FAILED

File: services/app-rs/src/enums.rs
/// Shared enums for the Rust URL shortener services.
///
/// These enums provide type safety for status fields across the codebase.
/// They serialize to strings for JSON compatibility with the Python stack.
use serde::{Deserialize, Serialize};

/// Health check status values.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum HealthStatus {
    Healthy,
    Unhealthy,
}

impl HealthStatus {
    /// Safely parse from string, falling back to Unhealthy for unknown values.
    pub fn from_str(s: &str) -> Self {
        match s {
            "healthy" => Self::Healthy,
            _ => Self::Unhealthy,
        }
    }
}

/// Service operational status values.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ServiceStatus {
    Pending,
    Running,
    Completed,
    Failed,
}

impl ServiceStatus {
    /// Safely parse from string, falling back to Failed for unknown values.
    pub fn from_str(s: &str) -> Self {
        match s {
            "pending" => Self::Pending,
            "running" => Self::Running,
            "completed" => Self::Completed,
            _ => Self::Failed,
        }
    }
}
